### ISSUE TITLE
Propagate transform changes from Node3D through to all children, even non-Node3D nodes

### ISSUE THREAD

**Issue by Awarets** [godotengine/godot-proposals#9122]
### Describe the project you are working on

A 3D FPS game

### Describe the problem or limitation you are having in your project

Transform changes from any given Node3D only propagate through to children if they are also Node3D.

When using a node setup such as (Node3D -> Node -> Node3D), transform changes will not propagate at all, and the children will instead behave as though the "Top Level" option is on.

This is problematic with complex 3D Node hierarchy as it requires certain nodes to inherit from Node3D even if a modified transform is not desired for that particular node.

See:
![image](https://github.com/godotengine/godot-proposals/assets/8617121/a57a75d3-1fbd-4c08-8f9f-659c1861ea23)

Having LagCompensatedTestSource inherit transform from PlayerView is desired, as this means that the source of the player's weapon raycasts will be from the player's viewpoint.
However, having WeaponFire inherit from Node3D is not desired, as there is no need for WeaponFire to have a transform at all, since the 3D positional source of the weapon raycast is already fulfilled by LagCompensatedTestSource.

I don't know how common my experience is, but in my opinion, this existing behaviour is unintuitive. I only discovered that it worked this way through trial and error, and I initially assumed it was a bug. 

Admittedly, the "Introduction to 3D" tutorial does include this line:

> Node3Ds have a local transform, which is relative to the parent node (as long as the parent node is also of or inherits from the type Node3D)

Which is my failing for not fully understanding, but it could maybe stand to include some additional information to prevent this confusion.

Additionally, if preventing transform inheritance IS desired, this is already achievable using the "Top Level" option.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow transform changes from Node3D to propagate through to all children unless that child is a Node3D with "Top Level" explicitly set to On.

This would allow for simpler setups of complex 3D scenes with greater separation of concerns, as Nodes would only inherit from Node3D if absolutely required.

I also believe that this would make transform inheritance simpler overall, as the three existing modes of operation:

- Node3D inherits transform from immediate Node3D parent
- Node3D does not inherit any transform due to its immediate parent not being Node3D, despite there being a Node3D in its lineage
- Node3D does not inherit any transform as "Top Level" is on

Down to just two modes:

- Node3D inherits transform from its closest Node3D parent
- Node3D does not inherit any transform as "Top Level" is on

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When a Node3D is translated, all of its Node3D children/grandchildren will be translated accordingly, even if there is a non-Node3D intermediary between them.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can at times be worked around by changing intermediary nodes to inherit from Node3D, but this is only possible if they are a base Node.

### Is there a reason why this should be core and not an add-on in the asset library?

This affects core engine functionality, I am not sure if it would be possible to achieve this with an add-on.

**Comment by Mickeon** [comment]
Use [RemoteTransform3D](https://docs.godotengine.org/en/stable/classes/class_remotetransform3d.html). It does one thing, but it does it well without requiring a script.

**Comment by theraot** [comment]
Why does `LagCompensatedTestSource` have to be a child of `WeaponFire`? Is it so `WeaponFire` or some other node can find it? If so, export a node variable for it, and don't have it as a child, so it can be transformed with the parent without further indirection.

**Comment by Awarets** [comment]
> Why does `LagCompensatedTestSource` have to be a child of `WeaponFire`? Is it so `WeaponFire` or some other node can find it? If so, export a node variable for it, and don't have it as a child, so it can be transformed with the parent without further indirection.

You're right that that would work, `LagCompensatedTestSource` is a child of `WeaponFire` simply because compositionally, `WeaponFire` is dependent on `LagCompensatedTestSource` to function.

I guess I'm just confused about what the purpose is of transform inheritance working the way it does now.
The negative response to this proposal leads me to assume that people are happy with or even dependent upon it, but I'm curious as to what a setup that utilises this inheritance-breaking mechanic looks like.

**Comment by Mickeon** [comment]
First of all, I'm very glad that you are willing to learn and are legitimately curious as to why. I hope you do not take this at heart and understand there is no ill-intent behind the negative reaction.

The negative reception is mainly for performance concerns. Some nodes do not do anything for this notification. Meanwhile, **CanvasItem** (**Control** & **Node2D**) share the same **NOTIFICATION_TRANSFORM_CHANGED** as **Node3D**, which would cause them to try and update as well, often pointlessly.
If the notification propagates for **all** of these nodes, indefinitely, **until no child is left unnotified**... that's wasteful. On a really big project, It can get really, _really_ bad.

The other reason is that this would break consistency with other properties that "propagate": If the proposal is accepted, should Node2D's `visible` and `modulate` propagate across all nodes as well? Should CanvasItem's `visible` affect Node3D?
Not only would it be inefficient (as mentioned before), but it would make things less self-contained, as everything would depend on the top-most Node3D unless _specifically opted out of it_.

> I'm curious as to what a setup that utilises this inheritance-breaking mechanic looks like.

There's a whole lot more projects that do **not** utilize this setup, at least not intentionally. I did have a project that combines **Node2D** and **Node3D** together and I wouldn't expect them to mix, personally.

**Comment by Awarets** [comment]
No worries, no offense is taken. I'm coming at this purely from the perspective of a user with limited knowledge of the inner workings of the engine, so my understanding of what would work best will be limited, and your explanation is much appreciated.

The performance concerns are completely understandable, and if push comes to shove I am definitely also in the camp of favoring performance, especially for a use case as niche as this.

> The other reason is that this would break consistency with other properties that "propagate": If the proposal is accepted, should Node2D's `visible` and `modulate` propagate across all nodes as well? Should CanvasItem's `visible` affect Node3D?

If the performance and implementation concerns could be resolved (which I am not in any way suggesting would be easy, I imagine that implementing this proposal might still not be worth the productivity cost even if we were all in favour of it), then I would actually say yes, I think all of these should propagate.

Perhaps I'm using the engine in an unusual way, but I feel like every child being beholden to their parent in this way makes sense.
Personally I can't really think of a use case for parent being made invisible while wanting a child to remain unaffected, that seems to me like those two nodes should not have a parent-child relationship in the first place.

**Comment by zenorbi** [comment]
I just ran into the same issue when developing an interaction system. Grouping of interactors is just a logical `Node`, but it had a side-effect of disconnecting the transform hierarchy.

I do understand the performance concerns this proposal would introduce given the current notification architecture. I am just curious. **If** backwards compatibility **wouldn't** be an issue, would a system like this work:

- When a node enters the tree it walks up the parent hierarchy, trying to find a node that has inheritable property.
  - `Node3D` tries to find a `Node3D` for its `transform` hierarchy
  - `CanvasItem` tries to find a `CanvasItem` for its `transform`, `module`, and `visible` hierarchy
  - etc.
-  A separate hierarchy would mirror the inheritance within the classes.

As far as I can see, `Node3D` already maintains a separate `List<Node3D *> children` field that only contains the `Node3D` children. `CanvasItem` also has a similar `List<CanvasItem *> children_items` field.

These fields are written to on `NOTIFICATION_ENTER_TREE` and `NOTIFICATION_EXIT_TREE`, so the hierarchy is almost fully there. The only thing missing in my opinion is the search across multiple parents.

I am probably wrong and this is a gross oversimplification of the problem, but I am curious if such a system would still be problematic.
